cpp/include/cudf/lists/stream_compaction.hpp

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace cudf::lists {
+
+/**
+ * @brief Filters elements in each row of `input` LIST column using `boolean_mask`
+ * LIST of booleans as a mask.
+ *
+ * Given an input `LIST` column and a list-of-bools column, the function produces
+ * a new `LIST` column of the same type as `input`, where each element is copied
+ * from the input row *only* if the corresponding `boolean_mask` is non-null and `true`.
+ *
+ * E.g.
+ * @code{.pseudo}
+ * auto const input    = lcw<int32_t>{ {0,1,2}, {3,4}, {5,6,7}, {8,9} };
+ * auto const selector = lcw<bool>   { {0,1,1}, {1,0}, {1,1,1}, {0,0} };
+ * auto const results  = apply_boolean_mask(lists_column_view{input}, lists_column_view{selector});
+ * results             == { {1,2}, {3}, {5,6,7}, {} };
```

Comment:
I think referencing things like `lcw` will be more confusing than helpful given it's only relevant to someone familiar with writing libcudf unit tests. 

```suggestion
 * input: { {0,1,2}, {3,4}, {5,6,7}, {8,9} };
 * mask:  { {0,1,1}, {1,0}, {1,1,1}, {0,0} };
 * result:{ {1,2},   {3},   {5,6,7}, {} };
```

Reply:
Agreed. This is less abstruse.
Fixed.